### ISSUE TITLE
chans NMDA, GABAB and VGCC bugs

### ISSUE THREAD

**Issue by rcoreilly** [emer/axon#204]
This is to document significant bugs in NMDA and VGCC equations in chans package, that are now fixed.

# NMDA missing vbio in numerator

In [chans/nmda.go](chans/nmda.go], the `-vbio` term was missing in the numerator (was just 1.0):

```Go
// MgGFmVbio returns the NMDA conductance as a function of biological membrane potential
// based on Mg ion blocking
func (np *NMDAParams) MgGFmVbio(vbio float32) float32 {
	vbio += np.Voff
	if vbio >= 0 {
		return 0
	}
	return -vbio / (1.0 + np.MgFact*mat32.FastExp(-0.062*vbio))
}
```

The difference is plotted here, using `Gbar = 0.004` in fixed version, and `Gbar = 0.15, Voff = 5` in the old buggy version, which shows that the advantage of Voff was likely because it makes the curve fit the accurate one better.  As usual, the biologically accurate equations just seem to work better.  Likewise, the new default for Voff is 0.

![fig_nmda_bug_voff5](https://user-images.githubusercontent.com/9605819/234938684-ccd4c9a9-dfb0-4fa4-91e1-eec60b475e36.png)

# CaFmVbio -- Calcium influx equation had buggy interpolation around 0

Here's the corrected version:

```
// CaFmVbio returns the calcium current factor as a function of biological membrane
// potential -- this factor is needed for computing the calcium current * MgGFmV.
// This is the same function used in VGCC for their conductance factor.
func (np *NMDAParams) CaFmVbio(vbio float32) float32 {
	vbio += np.Voff
	if vbio > -0.5 && vbio < 0.5 {
		return 1.0 / (0.0756 * (1 + 0.0378*vbio))
	}
	return -vbio / (1.0 - mat32.FastExp(0.0756*vbio))
}
```

And the comment in the code:
```
// Urakubo 2008 implementation in genesis
// http://kurodalab.bs.s.u-tokyo.ac.jp/info/STDP/ has this:
// if ({ abs {v} } < 0.5)
// max = {-1 / 0.0756 * {1 - 0.0378 * {v}}}
// if (V > -0.1 & V < 0.1){
//	  channel->Vca = -1/0.0756 + 0.5*V;
// }
// this was initially mis-coded as this without the minus sign:
// if vbio > -0.1 && vbio < 0.1 {
// 	return 1.0 / (0.0756 + 0.5*vbio)
// }
```

This linear approximation around 0 makes the function much smoother, if you happen to code it the right way.

The error was not evident because the nmda_plot and vgcc_plot functions weren't sampling at sufficiently tight step size to see it!!  doh.

**Comment by rcoreilly** [comment]
Unbelievable: there was also the same kind of bug in the chans GABAB implementation, like in NMDA, where the numerator was just plain missing.  

Here are the diffs for the record:

https://github.com/emer/axon/pull/193/files?diff=split&w=0#diff-78e85b013a36ef0e604e17434997e1755df048772c29f310db53e3030b9472ef

# key params changes:

These need to be more extensively tested, but preliminarily:

* NMDA.Gbar = .004 instead of .15.  .008 seems best for deep_fsa CT layer.
* GABAB.Gbar = .005 instead of .2.

**Comment by siboehm** [comment]
Can you do this in a separate PR? It doesn't really have anything to do with Novelty, would allow better PR reviews and neatly separate the changes.

**Comment by rcoreilly** [comment]
I can split at commit d408ed9291404f7e3e7c388030e43a278929703b -- the changes thereafter include param changes that depend on #202 and #205 that would not be productive to recapitulate, but that should capture most of the differences.

**Comment by siboehm** [comment]
sounds good!

**Comment by rcoreilly** [comment]
Everything working well now -- ra25 is significantly faster but everything else is pretty much the same -- surprising given how significant the differences are especially for GABA-B.  models are somehow too robust.  Fix mostly in #206.
